spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,526 @@
+#
+# Let's perform an [`error_analysis`](https://snorkel.readthedocs.io/en/redux/source/snorkel.analysis.html#module-snorkel.analysis.error_analysis) to see where our model makes mistakes.
+# We'll collect the predictions from the model and visualize examples in specific error buckets.
+
+# %%
+from snorkel.analysis.error_analysis import get_label_buckets
+from snorkel.analysis.utils import probs_to_preds
+
+outputs = model.predict(dl_valid, return_preds=True)
+error_buckets = get_label_buckets(
+    outputs["golds"]["spam_task"], outputs["preds"]["spam_task"]
+)
+
+# %% [markdown]
+# For application purposes, we might care especially about false negatives (true label was `1`, but model predicted `0`) — for the spam task, external links might point to malware, and we don't want to expose our users to these risks!
+
+# %%
+df_valid[["text", "label"]].iloc[error_buckets[(1, 0)]].head()
+
+# %% [markdown]
+# We notice that we're mis-classifying particularly some comments with shortened urls (e.g. `bit.ly/...`) — these links could redirect us to potentially dangerous websites, and we don't want our users to click them!
+
+# %% [markdown]
+# ## 3. Monitor data slices
+#
+# We leverage *slicing functions* (SFs) — an abstraction that shares syntax with *labeling functions*, which you should already be familiar with! (If not, please see the [intro tutorial](https://github.com/snorkel-team/snorkel-tutorials/blob/master/spam/01_spam_tutorial.ipynb).) A key difference: whereas labeling functions output labels, slicing functions output binary _masks_ indicating whether an example is in the slice or not.
+
+# %% [markdown]
+# In the following cells, we define a slicing function that identifies these shortened links the spam dataset.
+# To do so, we write a regex that checks for the commonly-used `.ly` extension.
+#
+# You'll notice that the slicing function is noisily defined — SFs are often heuristics to quickly measure performance over important subsets of the data.
+
+# %%
+import re
+from snorkel.slicing.sf import slicing_function
+
+
+@slicing_function()
+def short_link(x):
+    """Spam comments post links to other channels."""
+    return bool(re.search(r"\w+\.ly", x.text))
+
+
+sfs = [short_link]
+slice_names = [sf.name for sf in sfs]
+
+# %% [markdown]
+# For our $n$ examples and $k$ slices in each split, we apply the SF to our data to create an $n \times k$ matrix. (So far, $k=1$).
+
+# %%
+from snorkel.slicing.apply import PandasSFApplier
+
+applier = PandasSFApplier(sfs)
+S_train = applier.apply(df_train)
+S_valid = applier.apply(df_valid)
+S_test = applier.apply(df_test)
+
+# %% [markdown]
+# With a utility function from `snorkel.slicing.monitor`, we can visualize examples belonging to this slice in a `pandas.DataFrame`.
+
+# %%
+from snorkel.slicing.monitor import PandasSlicer
+
+pd_slicer = PandasSlicer(df_valid, sfs, slice_names)
+short_link_df = pd_slicer.slice("short_link")
```

Comment:
Should this be compositional?

Reply:
Now also thinking it's a bit odd to really only reference the SFs by name. Maybe it should be

```
pd_slicer = PandasSlicer(df_valid)
pd_slicer.slice(my_sf)
```

Reply:
nice!